ocaml/xapi/xapi_pif.ml

```diff
@@ -767,6 +767,31 @@ let set_property ~__context ~self ~name ~value =
         Nm.bring_pif_up ~__context pif
     ) (self :: vlan_pifs)
 
+let pif_has_clusters ~__context (self : API.ref_PIF) =
+  debug "Looking up clusters for PIF: %s" (Ref.string_of self);
+  let network_of_pif =
+    try Db.PIF.get_network ~__context ~self with _ ->
+      failwith (Printf.sprintf "Could not find network of PIF: %s" (Ref.string_of self))
+  in
+  info "Got network: %s" (Ref.string_of network_of_pif);
+  (Db.Cluster.get_records_where ~__context
+    ~expr:Db_filter_types.(Eq(Literal (Ref.string_of network_of_pif),Field "network")))
+    |> function
+    | []       -> false
+    | (a,_)::_ -> true
+
+let assert_no_clustering_enabled ~__context ~self =
+  if pif_has_clusters ~__context self
+  then raise Api_errors.(Server_error(Api_errors.cannot_change_pif_properties, [Ref.string_of self]))
+
+let set_disallow_unplug ~__context ~self ~value =
+  assert_no_clustering_enabled ~__context ~self;
+  if Db.PIF.get_disallow_unplug ~__context ~self <> value
+  then begin
+    info "PIF.set_disallow_unplug: PIF: %s, value: %s" (Ref.string_of self) (string_of_bool value);
```

Comment:
I would remove this debug line, since it's almost the same as the one in message_forwarding.

Reply:
Also when you add a debug message it would be nice if it printed UUIDs instead of OpaqueRefs, it makes debugging easier with `xe`. I've fixed that in a few places in the clustering code.
Don't fix it on this one since you're dropping it anyway, but in the future.

Reply:
I think that's not a huge problem if we have the database dump, because then we can match UUIDs to opaquerefs.